toolchain/check/handle_where.cpp

```diff
@@ -8,41 +8,78 @@
+  auto self_type_id = ExprAsType(context, self_node, self_id);
+  // TODO: Do this instead once `WhereExpr` is ready to consume this:
+  // context.node_stack().Push(node_id, self_type_id);
+  context.node_stack().Push(self_node, self_id);
 
-auto HandleParseNode(Context& context, Parse::WhereOperandId node_id) -> bool {
-  return context.TODO(node_id, "HandleWhereOperand");
+  // Introduce a name scope so that we can remove the `.Self` entry we are
+  // adding to name lookup at the end of the `where` expression.
+  context.scope_stack().Push();
+  // Introduce `.Self` as a symbolic binding. Its type is the value of the
+  // expression to the left of `where`, so `MyInterface` in the example above.
+  // Because there is no equivalent non-symbolic value, we use `Invalid` as
+  // the `value_id` on the `BindSymbolicName`.
+  auto entity_name_id = context.entity_names().Add(
+      {.name_id = SemIR::NameId::PeriodSelf,
+       .parent_scope_id = context.decl_name_stack().PeekParentScopeId(),
+       .bind_index = context.scope_stack().AddCompileTimeBinding()});
+  // FIXME: should this have a location associated with `node_id`?
+  auto inst_id =
+      context.AddInst(SemIR::LocIdAndInst::NoLoc<SemIR::BindSymbolicName>(
+          {.type_id = self_type_id,
+           .entity_name_id = entity_name_id,
+           .value_id = SemIR::InstId::Invalid}));
+  context.scope_stack().PushCompileTimeBinding(inst_id);
+  auto existing =
+      context.scope_stack().LookupOrAddName(SemIR::NameId::PeriodSelf, inst_id);
+  // Shouldn't have any names in newly created scope.
+  CARBON_CHECK(!existing.is_valid());
+  return true;
 }
 
-auto HandleParseNode(Context& context, Parse::RequirementEqualId node_id)
+auto HandleParseNode(Context& context, Parse::RequirementEqualId /*node_id*/)
     -> bool {
-  return context.TODO(node_id, "HandleRequirementEqual");
+  // TODO: Implement
+  context.node_stack().PopExpr();
+  context.node_stack().PopExpr();
+  return true;
 }
 
-auto HandleParseNode(Context& context, Parse::RequirementEqualEqualId node_id)
-    -> bool {
-  return context.TODO(node_id, "HandleRequirementEqualEqual");
+auto HandleParseNode(Context& context,
+                     Parse::RequirementEqualEqualId /*node_id*/) -> bool {
+  // TODO: Implement
+  context.node_stack().PopExpr();
+  context.node_stack().PopExpr();
+  return true;
 }
 
-auto HandleParseNode(Context& context, Parse::RequirementImplsId node_id)
+auto HandleParseNode(Context& context, Parse::RequirementImplsId /*node_id*/)
     -> bool {
-  return context.TODO(node_id, "HandleRequirementImpls");
+  // TODO: Implement
+  context.node_stack().PopExpr();
+  context.node_stack().PopExpr();
+  return true;
 }
 
-auto HandleParseNode(Context& context, Parse::RequirementAndId node_id)
+auto HandleParseNode(Context& /*context*/, Parse::RequirementAndId /*node_id*/)
     -> bool {
-  return context.TODO(node_id, "HandleRequirementAnd");
+  // Nothing to do
```

Comment:
```suggestion
  // Nothing to do.
```

[Punctuation style](https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar)